### PR TITLE
Performance Improvements

### PR DESCRIPTION
**Pull request by inteon**
Greatly improves performance of the go-licenses tool by preventing a duplicate expensive operations and parallelising them.

The performance improvement for cert-manager:
```console
# master
$ time make update-licenses
...
real    2m12.426s
user    1m0.442s
sys     0m2.430s
```

```console
# this PR
$ time make update-licenses
...
real    0m13.249s
user    0m17.688s
sys     0m1.853s
```

### CODE REVIEW

File: licenses/classifier.go [hunk 1]
```diff
@@ -63,21 +62,25 @@ func (c *googleClassifier) Identify(licensePath string) ([]License, error) {
...
 
+	foundLicenseNames := map[string]struct{}{}
+
 	licenses := []License{}
 	for _, match := range matches.Matches {
 		if match.MatchType != "License" {
 			continue
 		}
 
+		// Skip duplicate licenses.
```
Comment (inteon) [inline licenses/classifier.go:73]: This is an unrelated improvement: if the same license is found in a LICENSE file twice, only display it once.

**Comment by inteon** [comment]
@Bobgy this PR is ready for review.

File: check.go [hunk 2]
```diff
@@ -81,18 +81,13 @@ func checkMain(_ *cobra.Command, args []string) error {
 	found := false
 
 	for _, lib := range libs {
-		if lib.LicensePath == "" {
+		if lib.LicenseFile == "" {
```
Comment (inteon) [inline check.go:84]: I renamed `LicensePath` to `LicenseFile` so it is clear that we are not talking about a go path here.

**Comment by inteon** [comment]
@Bobgy Is there something I can do still to get this merged?

**Comment by Bobgy** [comment]
> @Bobgy Is there something I can do still to get this merged?

Sorry I am not spending enough time on this. I tried looking at it initially, but some part of the code was a bit hard to understand. I don't remember now.

I think the simplest way to give us enough confidence is by setting up a bigger e2e test example and confirm the result has no diffence for this PR.

**Comment by inteon** [comment]
@Bobgy I added a more complex e2e test in https://github.com/google/go-licenses/pull/215.
This PR also runs this test successfully, giving us more confidence in this PR.

**Comment by Bobgy** [comment]
Thank you! I will take another quick look

**Review by Bobgy — COMMENTED** [review]
Generally looks great! I need to read library.go again not on my phone.

File: licenses/find.go [hunk 3]
```diff
@@ -27,69 +26,57 @@ var (
 	licenseRegexp = regexp.MustCompile(`^(?i)((UN)?LICEN(S|C)E|COPYING|README|NOTICE).*$`)
 )
 
-// Find returns the file path of the license for this package.
+// FindCandidates returns the candidate file path of the license for this package.
```
[the remark this answers is not in the document]
Comment (inteon) [reply licenses/find.go:29]: This way, we can find all candidates and deduplicate them & only classify each candidate once across all `pkgDir`s.

File: licenses/module.go [hunk 4]
```diff
@@ -33,9 +33,6 @@ type Module struct {
 }
 
 func newModule(mod *packages.Module) *Module {
-	if mod == nil {
```
[the remark this answers is not in the document]
Comment (inteon) [reply licenses/module.go:36]: For both locations where this function is used, the argument will not be nil.
Also, in both locations, the output is assumed to be non-nil too.

File: licenses/find.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (inteon) [reply licenses/find.go:29]: The Libraries has been reworked and works as follows now:
1. we list all modules and all packages
2. for each package, we find a list of candidates
3. we deduplicate all candidates
4. for each candidate, we classify if the candidate is a license file
5. for each package, we select the first candidates that is a license file & add the package to a list of packages for that license file
6. we return an array of libraries (which are the license files, the found licenses in that file, all the packages that had that file as its first candidate and the module in which those packages live)

[the remark this answers is not in the document]
Comment (inteon) [reply licenses/find.go:29]: I added the comment here: https://github.com/google/go-licenses/pull/204/commits/e3146eb3fa59542a46571d7a252cce0d70b59510
@Bobgy

**Comment by inteon** [comment]
@Bobgy Is this PR ready to be merged?

**Comment by inteon** [comment]
Could you verify that the different license `github.com/jmespath/go-jmespath` change also happens on master? I think this is due to the new classifier version that we are using.

**Comment by gwatts** [comment]
I just created a new throwaway Go project here with a package that imports `github.com/jmespath/go-jmespath` and `github.com/gwatts/rootcerts`

With `go-license` 1.6.0 it outputs (using `go-licenses report ./... --ignore github.com/gwatts/lictest`:

```csv
github.com/gwatts/rootcerts,https://github.com/gwatts/rootcerts/blob/ff6b9d3bf583/LICENSE,MIT
github.com/jmespath/go-jmespath,https://github.com/jmespath/go-jmespath/blob/v0.4.0/LICENSE,Apache-2.0
```

As of commit e41525ccefa8f75f4b790debd8a674ede5b49190 though it doesn't output `go-jmespath` at all for this test case:

```
github.com/gwatts/rootcerts,https://github.com/gwatts/rootcerts/blob/ff6b9d3bf583/LICENSE,MIT
```

And for this PR branch:
```
E0825 19:37:39.480783 1029605 report.go:143] Error identifying license in "Unknown": no license found
github.com/gwatts/rootcerts,https://github.com/gwatts/rootcerts/blob/ff6b9d3bf583/LICENSE,MIT
github.com/jmespath/go-jmespath,Unknown,Unknown
```

**Comment by inteon** [comment]
@gwatts https://github.com/jmespath/go-jmespath/blob/v0.4.0/LICENSE not being recoginised is due to a change in the behavior of licenseclassifier/v2 vs licenseclassifier, and I don't think we can do much about it. It also looks like the particular LICENSE file has been updated to a more uniform format: https://github.com/jmespath/go-jmespath/commit/b89213a6ec69212a244ef0d07d893bef7f2d9091

Thanks to pointing out that the `Unknown,Unknown` entry is missing on master right now. Also, there is no error on master.
This means that this code is not executed on master: https://github.com/google/go-licenses/pull/204/files#diff-fee12aa2a5b1e9be6ab94d44ddebc2acd1d3a79d5183340ceee62dd114d49157L112
But it is executed in this branch: https://github.com/google/go-licenses/pull/204/files#diff-fee12aa2a5b1e9be6ab94d44ddebc2acd1d3a79d5183340ceee62dd114d49157R142-R150

I think we should actually show an error and add the Unknown entry because we did not find a license. We should probably improve the error message (`in "Unknown"` does not make sense).

**Review by Bobgy — APPROVED** [review]

**Comment by Bobgy** [comment]
Thank you @inteon for the awesome perf improvements! It's great to see the test numbers.

Let's merge and release v2 alpha first.
